File: docs/public/code/exampleAgentFramework/15.Langchain Agent/node/agent.ts
/**
 * 最小 createAgent：一件模拟天气工具。
 */
import { createAgent, tool } from 'langchain'
import { z } from 'zod'

import { createChatModel } from './client'

const getWeather = tool(
  async ({ city }) => `${city}：晴，26°C`,
  {
    name: 'get_weather',
    description: '查询指定城市的天气。',
    schema: z.object({
      city: z.string().describe('城市名，例如杭州'),
    }),
  },
)

export function createWeatherAgent() {
  return createAgent({
    model: createChatModel(),
    tools: [getWeather],
    systemPrompt: '你是助手。问天气就调用 get_weather，用中文回答。',
  })
}

export function lastText(messages: Array<{ content?: unknown }>): string {
  const last = messages.at(-1)
  const content = last?.content
  if (typeof content === 'string') {
    return content
  }
  if (Array.isArray(content)) {
    return content
      .map((part) => {
        if (typeof part === 'string') {
          return part
        }
        if (part && typeof part === 'object' && 'text' in part) {
          return String((part as { text?: string }).text ?? '')
        }
        return ''
      })
      .join('')
  }
  return JSON.stringify(content ?? '')
}

export async function askWeather(question: string): Promise<string> {
  const agent = createWeatherAgent()
  const result = await agent.invoke({
    messages: [{ role: 'user', content: question }],
  })
  return lastText(result.messages as Array<{ content?: unknown }>)
}

File: docs/public/code/exampleAgentFramework/15.Langchain Agent/python/agent.py
"""最小 create_agent：一件模拟天气工具。"""
from langchain.agents import create_agent

from client import create_chat_model


def get_weather(city: str) -> str:
    """查询指定城市的天气。"""
    return f"{city}：晴，26°C"


def create_weather_agent():
    return create_agent(
        model=create_chat_model(),
        tools=[get_weather],
        system_prompt="你是助手。问天气就调用 get_weather，用中文回答。",
    )


def ask_weather(question: str) -> str:
    agent = create_weather_agent()
    result = agent.invoke({"messages": [{"role": "user", "content": question}]})
    last = result["messages"][-1]
    content = getattr(last, "content", last)
    return content if isinstance(content, str) else str(content)
